### PR TITLE
Skip loading from instant execution cache on --refresh-dependencies

### PR DESCRIPTION
**Pull request by eskatos**

### CODE REVIEW

File: subprojects/instant-execution/src/integTest/groovy/org/gradle/instantexecution/InstantExecutionIntegrationTest.groovy [hunk 1]
```diff
@@ -104,6 +104,28 @@ class InstantExecutionIntegrationTest extends AbstractIntegrationSpec {
...
+        instantRun "help"
+
+        then:
+        outputDoesNotContain("classic")
+
+        when:
+        instantRun "help", "--refresh-dependencies"
+
+        then:
+        outputContains("classic")
```
Comment (bamboo) [inline subprojects/instant-execution/src/integTest/groovy/org/gradle/instantexecution/InstantExecutionIntegrationTest.groovy]: ⭕️ I think I would prefer a test that would actually change the cached state and assert that not only `--refresh-dependencies` causes build logic to run but that the cached state is then different after it

File: subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/DefaultInstantExecution.kt [hunk 2]
```diff
@@ -78,18 +81,24 @@ class DefaultInstantExecution(
...
-            return false
-        } else {
-            logger.lifecycle("Reusing instant execution cache. This is not guaranteed to work in any way.")
-            return true
+    override fun canExecuteInstantaneously(): Boolean =
+        when (isInstantExecutionEnabled) {
+            false -> false
+            true -> when {
+                host.isInvalidateCache -> {
+                    logger.lifecycle("Calculating task graph as instant execution cache invalidation was required")
```
Comment (bamboo) [inline subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/DefaultInstantExecution.kt]: :x: `... was requested"` ?

File: subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/InstantExecutionHost.kt [hunk 3]
```diff
@@ -85,18 +81,25 @@ class InstantExecutionHost internal constructor(
...
 
+    override val isInvalidateCache: Boolean
+        get() = gradle.startParameter.isRefreshDependencies
+
     override val currentBuild: ClassicModeBuild =
         DefaultClassicModeBuild()
 
     override fun createBuild(rootProjectName: String): InstantExecutionBuild =
-        DefaultInstantExecutionBuild(rootProjectName)
+        DefaultInstantExecutionBuild(getService(PathToFileResolver::class.java), rootProjectName)
```
Comment (bamboo) [inline subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/InstantExecutionHost.kt]: ⭕️ just a refactoring?

Comment (eskatos) [reply subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/InstantExecutionHost.kt]: yep, see https://github.com/gradle/gradle/pull/9409/commits/12eb5215e91ff039c58c9067b4eb42d703acf2a0

File: subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/DefaultInstantExecution.kt (hunk 2 above)
Comment (eskatos) [reply subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/DefaultInstantExecution.kt]: Fixed, thanks

File: subprojects/instant-execution/src/integTest/groovy/org/gradle/instantexecution/InstantExecutionIntegrationTest.groovy (hunk 1 above)
Comment (eskatos) [reply subprojects/instant-execution/src/integTest/groovy/org/gradle/instantexecution/InstantExecutionIntegrationTest.groovy]: It's too early in the spike to implement that as part of this pr, let's do it soon enough

**Comment by adammurdoch** [comment]
`--refresh-dependencies` can be extremely slow, so I'm not sure we should overload it even for a spike (particularly if we want to measure performance). I think we have some other options:

- Give `org.gradle.unsafe.instant-execution` a value: `e.g. enable`, `disable`, `refresh`
- Add some new command-line options, e.g. `--unsafe-instant-execution` and `--unsafe-refresh-instant-execution`
- Add another system property

**Comment by adammurdoch** [comment]
... or just document `rm -rf .instant-execution-state` for now

**Comment by eskatos** [comment]
Closing as dropped

**Comment by bamboo** [comment]
We can certainly introduce a faster, better way to invalidate the cache BUT it was still correct to do it during `--refresh-dependencies`. I wish you hadn't dropped this one.

**Comment by eskatos** [comment]
@bamboo reopened and reworked the code, please take another look

File: subprojects/instant-execution/src/integTest/groovy/org/gradle/instantexecution/InstantExecutionIntegrationTest.groovy (hunk 1 above)
Comment (eskatos) [reply subprojects/instant-execution/src/integTest/groovy/org/gradle/instantexecution/InstantExecutionIntegrationTest.groovy]: Proper test added, using a custom task type worked until custom task actions are supported

File: subprojects/instant-execution/src/integTest/groovy/org/gradle/instantexecution/InstantExecutionSkipCachedStateIntegrationTest.groovy [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        outputContains("foo")
+
+        when:
+        instantRun "myTask", "--refresh-dependencies"
+
+        then:
+        outputContains("bar")
+
+        when:
+        instantRun "myTask"
```
Comment (bamboo) [inline subprojects/instant-execution/src/integTest/groovy/org/gradle/instantexecution/InstantExecutionSkipCachedStateIntegrationTest.groovy]: ⭕️ Assert the cached work graph was reused here?

Comment (eskatos) [reply subprojects/instant-execution/src/integTest/groovy/org/gradle/instantexecution/InstantExecutionSkipCachedStateIntegrationTest.groovy]: What about we introduce build operations for saving and loading the work graph so we can assert on them, have them record timings and be shown in e.g. chrome traces? I'd do that in a subsequent PR though, reworking all existing tests to leverage them. How does that sound?

Comment (bamboo) [reply subprojects/instant-execution/src/integTest/groovy/org/gradle/instantexecution/InstantExecutionSkipCachedStateIntegrationTest.groovy]: Sounds great.

**Pull request merged**